File: ts/twoSum.ts
// Given an array of integers nums and an integer target
// return indices of the two numbers such that they add up to target.
function twoSum(nums: number[], target: number): number[] {

    // 1. Create hashmap for notebook
    let hashmap: Map<number, number> = new Map();

    // 2. Iterate the array
    for(let i=0; i < nums.length; i++) {
        // 3. Finds what's needed to reach target
        let diff = target - nums[i] ;
        
        // 4. Check if we already have the diff number in notebook
        if (hashmap.has(diff)) {

            // 5. Return the key
            let num1 = hashmap.get(diff);
            
            return [num1 as number, i];
        }
        // 6. Keep the number and indice in the notebook
        hashmap.set(nums[i], i);
    }
    return [];
}

let array: number[] = [10, 7, 3, 4, 2, 1];
console.log(twoSum(array, 9));

File: java/src/main/java/TwoSum.java
import java.util.Arrays;
import java.util.HashMap;

public class TwoSum {
    public static void main(String[] args) {
        int[] nums = {2, 7, 11, 15};
        int target = 9;

        // expected output: [0, 1]
        System.out.println(Arrays.toString(twoSum(nums, target)));

    }

    public static int[] twoSum(int[] nums, int target) {
        // Memory
        HashMap<Integer, Integer> hashMap = new HashMap<>();

        // Iterate
        for(int i=0; i < nums.length; i++) {
            // Complement = 9 - current num
            int complement = target - nums[i];
            
            // do we already have the complement in the memory?
            if (hashMap.containsKey(complement)) {
                // Return
                return new int[] { hashMap.get(complement), i};
            } else {
                // Keep tracking of value & its index
                hashMap.put(nums[i], i);
            }
        }
        return new int[] {};
    }
}
